src/cache.h

```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
+ * cache_reset_stats
+ *
+ * Debugging facility.
+ * Resets performance statistics counters.
+ *-----------------------------------------------------------------------------
+ */
 static inline void
 cache_reset_stats(cache *cc)
 {
    return cc->ops->reset_stats(cc);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_reset_stats
+ *
+ * Debugging facility.
+ * Returns performance statistics counts.
+ *-----------------------------------------------------------------------------
+ */
 static inline void
 cache_io_stats(cache *cc, uint64 *read_bytes, uint64 *write_bytes)
 {
    return cc->ops->io_stats(cc, read_bytes, write_bytes);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_page_valid
+ *
+ * Debugging facility.
+ * Returns false if addr is unaligned or out-of-bounds, or if addr has no get
+ * references (presumably a call-graph ancestor caller holds at least one).
+ *-----------------------------------------------------------------------------
+ */
 static inline bool
 cache_page_valid(cache *cc, uint64 addr)
 {
    return cc->ops->page_valid(cc, addr);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_validate_page
+ *
+ * Debugging facility.
+ * Asserts cache_page_valid, that the addr matches what page points to,
+ * and that the page isn't free.
+ *-----------------------------------------------------------------------------
+ */
 static inline void
 cache_validate_page(cache *cc, page_handle *page, uint64 addr)
 {
    cc->ops->validate_page(cc, page, addr);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_count_dirty
+ *
+ * Debugging facility.
+ * Returns the number of dirty pages in the cache.
+ *-----------------------------------------------------------------------------
+ */
 static inline uint32
 cache_count_dirty(cache *cc)
 {
    return cc->ops->count_dirty(cc);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_get_read_ref
+ *
+ * Debugging facility.
+ * Returns the number of threads with references to page.
+ *-----------------------------------------------------------------------------
+ */
 static inline uint32
 cache_get_read_ref(cache *cc, page_handle *page)
 {
    return cc->ops->page_get_read_ref(cc, page);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_present
+ *
+ * Debugging facility.
+ * Returns TRUE if page is present in the cache.
+ *-----------------------------------------------------------------------------
+ */
 static inline bool
 cache_present(cache *cc, page_handle *page)
 {
    return cc->ops->cache_present(cc, page);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_enable_sync_get
+ *
+ * XXX No idea.
```

Comment:
This is a safety function. There is a known deadlock with async gets, so they are (were?) disabled by default. This function must be called to enable their use or perhaps an assertion is triggered.

Reply:
Wait, this function enables *sync* gets, not async gets?

Reply:
Is the known deadlock between a sync get and an async get, or between any two async gets? Or something else?